### PR TITLE
Added some coverage for executors and test cases for secret mounting and init-containers

### PR DESCRIPTION
**Pull request by liyinan926**
@kimoonkim @foxish

### CODE REVIEW

**Comment by foxish** [comment]
@liyinan926 given the recently uncovered bug, I'm curious - did these tests detect that issue?

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageBuilder.scala
Comment (foxish) [inline integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/docker/SparkDockerImageBuilder.scala]: Please also add an option to allow the test to directly use an init container supplied externally from outside the test code, similar to the options for the driver and executor.

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (foxish) [inline integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala]: typo meant to be `etc`?

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (foxish) [inline integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:264]: Is this option being exercised currently?

**Review by foxish — COMMENTED** [review]
Mostly LGTM - just a couple of clarifications.

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (liyinan926) [reply integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:264]: No, none of the tests uses an init-container yet.

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (liyinan926) [reply integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala]: Yeah, fixed.

**Comment by liyinan926** [comment]
@foxish the current set of tests won't uncover the bug because none of them really trigger the use of an init-container. We need tests that use a remote dependency so an init-container is added and used.

**Review by kimoonkim — COMMENTED** [review]
This looks good to me overall, pending @foxish's comments.

Did you test this by running it locally? I wonder if we want to have a Jenkins job trigger on PRs to this repo.

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
A Jenkins run has passed with this PR, http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-dev-spark-integration/2/console.

But it failed to post a status here. Mabye @foxish can give the Jenkins account the write access to this repo?
 
> Setting status of 6a1353fd0343926578f713d2e824e67e121c619a to SUCCESS with url http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-dev-spark-integration/2/ and message: 'Build completed successfully.
 '
Using context: Integration Tests
FileNotFoundException means that the credentials Jenkins is using is probably wrong. Or the user account does not have write access to the repo.
java.io.FileNotFoundException: {"message":"Not Found","documentation_url":"https://developer.github.com/v3/repos/statuses/#create-a-status"}

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (kimoonkim) [inline integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:84]: Hmm. It seems appArgs are ignored in the spark-submit command line. Please see https://github.com/apache-spark-on-k8s/spark-integration/blob/master/integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesTestComponents.scala#L107

Comment (liyinan926) [reply integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:84]: Good catch! Fixed.

**Comment by kimoonkim** [comment]
rerun integration tests please

**Comment by liyinan926** [comment]
@kimoonkim @foxish I just added a couple of test cases that use a remote example jar hosted at https://storage.googleapis.com/spark-k8s-integration-tests/jars/spark-examples_2.11-2.3.0.jar. The tests will trigger the use of the init-container. See https://github.com/apache-spark-on-k8s/spark-integration/pull/15/commits/e2d20c93d4d95caddf1dc80f76cc60f06bb3bbab.

**Comment by kimoonkim** [comment]
Jenkins job is now running on this PR and commenting on it!

**Comment by foxish** [comment]
Awesome.. thanks Kimoon & Yinan! Will take another look shortly.

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (foxish) [inline integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala]: Don't we call this `spark-init` elsewhere?

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (foxish) [inline integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:137]: Is this test using the remote example jar? I can't find it being referenced in the test itself.

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (liyinan926) [reply integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala]: Yes, renamed to `spark-init`.

File: integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (liyinan926) [reply integration-test/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:137]: Fixed.

**Comment by liyinan926** [comment]
rerun integration tests please

**Comment by kimoonkim** [comment]
The latest Jenkins run failed with "spark-base/entrypoint.sh: no such file or directory":
> *** RUN ABORTED ***
  com.spotify.docker.client.exceptions.DockerException: ProgressMessage{id=null, status=null, stream=null, error=lstat spark-base/entrypoint.sh: no such file or directory, progress=null, progressDetail=null}
  at com.spotify.docker.client.LoggingBuildHandler.progress(LoggingBuildHandler.java:33)
  at com.spotify.docker.client.DefaultDockerClient.build(DefaultDockerClient.java:1157)
  at org.apache.spark.deploy.k8s.integrationtest.docker.SparkDockerImageBuilder.buildImage(SparkDockerImageBuilder.scala:70)
  at org.apache.spark.deploy.k8s.integrationtest.docker.SparkDockerImageBuilder.buildSparkDockerImages(SparkDockerImageBuilder.scala:64)
  at org.apache.spark.deploy.k8s.integrationtest.backend.minikube.MinikubeTestBackend.initialize(MinikubeTestBackend.scala:31)
  at org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.beforeAll(KubernetesSuite.scala:44)
  at org.scalatest.BeforeAndAfterAll$class.beforeAll(BeforeAndAfterAll.scala:187)
  at org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.beforeAll(KubernetesSuite.scala:35)
  at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:253)
  at org.apache.spark.deploy.k8s.integrationtest.KubernetesSuite.org$scalatest$BeforeAndAfter$$super$run(KubernetesSuite.scala:35)
  ...

**Comment by foxish** [comment]
Ah! This might be because the dockerfiles have changed in https://github.com/apache/spark/pull/20154

**Comment by liyinan926** [comment]
Oh, yes, it now requires `ARG img_path`.

**Comment by foxish** [comment]
I'm not sure how the builder plugin does this. I'm starting to think that since we'll be using the build-push-docker-images script to create release-images also in future, we should just invoke that to build images even for minikube. The script could be an e2e-minikube.sh, similar to e2e-cloud.sh in https://github.com/apache-spark-on-k8s/spark-integration/pull/16/

**Comment by liyinan926** [comment]
I'm investigating if I can add build arguments with the docker client.

**Comment by liyinan926** [comment]
@foxish @kimoonkim It doesn't seem to be picking up my docker build argument changes at https://github.com/apache-spark-on-k8s/spark-integration/pull/15/commits/5decbce9c93e1beb3602b856368ab55bce4a9b9c. The build seemed still using the code already in the repo.

**Review by foxish — APPROVED** [review]
LGTM, pending a manual/automated run of the tests themselves.

**Comment by foxish** [comment]
@liyinan926, can you also add the init-container image option to the e2e/runner.py's invocation of the test, so that we supply the init-container to the test code, similar to https://github.com/apache-spark-on-k8s/spark-integration/blob/master/e2e/runner.sh#L115

**Comment by liyinan926** [comment]
@foxish done.

**Review by foxish — APPROVED** [review]

**Comment by foxish** [comment]
Merging. Thanks!

**Comment by liyinan926** [comment]
Manually ran the tests locally and all passed.

**Pull request merged**
